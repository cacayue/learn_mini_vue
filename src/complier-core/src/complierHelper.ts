export const TO_DISPLAY_STRING = Symbol('toDisplayString');

export const helperNameMap = new Map([[TO_DISPLAY_STRING, 'toDisplayString']]);
