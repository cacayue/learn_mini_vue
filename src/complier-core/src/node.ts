export const enum NodeType {
  INTERPOLATION,
  SIMPLE_EXPRESSION
}

export type ParseContext = {
  source: any;
};
