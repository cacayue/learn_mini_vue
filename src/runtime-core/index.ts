export { createRender } from './render';
export { h } from './h';
export { renderSlot } from './helper/renderSlots';
export { createTextVNode } from './vnode';
export { getCurrentInstance } from './component';
export { provide, inject } from './apiInject';
export { nextTick } from './scheduler';
