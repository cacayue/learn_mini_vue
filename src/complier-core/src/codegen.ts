import { helperNameMap, TO_DISPLAY_STRING } from './complierHelper';
import { NodeType } from './node';

export function generateCode(ast: any) {
  let context = createContext();
  const { push } = context;
  genFunctionPreamble(ast, context);
  const functionName = 'render';
  const args = ['_ctx', '_cache'];
  const signature = args.join(', ');
  const node = ast.codeGenNode;
  push(`function ${functionName}(${signature}) {`);

  genNode(node, context);

  push(`}`);
  return {
    code: context.code
  };
}

function genFunctionPreamble(ast: any, context: any) {
  const { push } = context;
  const VueBinging = `'vue'`;

  if (ast.helpers?.length > 0) {
    const aliasHelper = (s: string) => `${s} as _${s}`;
    push(`import { ${ast.helpers.map(aliasHelper)} } from ${VueBinging}`);
    push('\n');
  }

  push('return ');
}

function createContext() {
  const context = {
    code: '',
    push(source: string) {
      context.code += source;
    },
    helper(key: symbol) {
      return `_${helperNameMap.get(key)}`;
    }
  };
  return context;
}

function genNode(node: any, context: any) {
  const { push } = context;
  push(`return `);
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context);
      break;
    case NodeType.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    default:
      break;
  }
}

function genText(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context;
  console.log(node.content);
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(`)`);
}
function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}
