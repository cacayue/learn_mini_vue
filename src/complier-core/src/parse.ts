import { NodeType, ParseContext } from './node';

const openDelimiter = '{{';
const closeDelimiter = '}}';
const startTag = '<';

const enum TagType {
  Start,
  End
}

export function baseParse(content: string) {
  const context = createParseContext(content);

  return createRoot(parseChildren(context));
}

function parseChildren(context: ParseContext) {
  let nodes = [];

  while (!isEnd(context)) {
    let node: any | undefined = undefined;
    let s = context.source.trim();
    if (s.startsWith(openDelimiter)) {
      node = parseInterpolation(context);
    } else if (s[0] === startTag) {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context);
      }
    }

    if (!node) {
      node = parseText(context);
    }

    nodes.push(node);
  }

  function isEnd(context: ParseContext) {
    context.source = context.source.trim();
    if (context.source.startsWith('</div>')) {
      return true;
    }

    return !context.source;
  }

  return nodes;
}

function parseText(context: ParseContext) {
  let endIndex = context.source.length;
  let endTokenIndex = context.source.indexOf(openDelimiter);

  if (endTokenIndex !== -1) {
    endIndex = endTokenIndex;
  }

  // 1. 获取content
  const content = parseTextData(context, endIndex);

  return {
    type: NodeType.TEXT,
    content
  };
}

function parseTextData(context: ParseContext, length: number) {
  const content = context.source.slice(0, length);

  // 2. 推进
  advanceBy(context, content.length);
  return content;
}

function parseElement(context: ParseContext) {
  context.source = context.source.trim();
  // <div></div>
  // 解析开始tag
  // 删除解析过的
  let element: any = parseTag(context, TagType.Start);
  element.children = parseChildren(context);

  context.source = context.source.trim();

  // 再次解析结束tag
  // 删除解析过的
  parseTag(context, TagType.End);

  return element;
}

function parseTag(context: ParseContext, tagType: TagType) {
  let match = /^<\/?([a-z]*)/i.exec(context.source);
  if (!match) {
    return undefined;
  }

  let tag = match[1];
  advanceBy(context, match[0].length);
  advanceBy(context, 1);

  if (tagType === TagType.End) {
    return;
  }

  return {
    type: NodeType.ELEMENT,
    tag: tag
  };
}

function parseInterpolation(context: ParseContext) {
  // {{message}}

  // 9
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length);

  // message}}
  advanceBy(context, openDelimiter.length);

  // 7
  const rawContentLength = closeIndex - closeDelimiter.length;

  // 截取0-7：  message
  const rawContent = parseTextData(context, rawContentLength);

  // 推进
  advanceBy(context, closeDelimiter.length);
  const content = rawContent.trim();

  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content: content
    }
  };
}

// 推进： 每次处理完成对应的字符后截取掉，只保留没处理的
function advanceBy(context: ParseContext, length: number) {
  context.source = context.source.slice(length);
}

function createParseContext(content: string): ParseContext {
  return {
    source: content
  };
}

function createRoot(children: any[]) {
  return {
    children: children
  };
}
