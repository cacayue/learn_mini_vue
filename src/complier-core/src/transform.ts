import { helperNameMap, TO_DISPLAY_STRING } from './complierHelper';
import { NodeType } from './node';

export type TransformOptions = {
  nodeTransforms: Array<Function>;
};

export type TransformContext = {
  root: any;
  nodeTransforms: Array<Function>;
  helpers: Array<string>;
  helper: Function;
};

export function transform(root: any, options?: TransformOptions) {
  const context = createTraverseContext(root, options);
  traverseNode(root, context);

  createRootCodegen(root);

  root.helpers = [...context.helpers];
}

function createRootCodegen(root: any) {
  root.codeGenNode = root.children[0];
}

function traverseNode(node: any, context: TransformContext) {
  const transforms = context.nodeTransforms;

  for (let i = 0; i < transforms.length; i++) {
    const transform = transforms[i];
    transform(node);
  }

  switch (node.type) {
    case NodeType.INTERPOLATION:
      context.helper(helperNameMap.get(TO_DISPLAY_STRING));
      break;
    case NodeType.ROOT:
    case NodeType.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
}

// 1. 广度优先搜索 遍历AST
function traverseChildren(node: any, context: TransformContext) {
  let children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      traverseNode(child, context);
    }
  }
}

function createTraverseContext(root: any, options?: TransformOptions): TransformContext {
  let context: TransformContext = {
    root,
    nodeTransforms: options?.nodeTransforms || [],
    helpers: [],
    helper: (s: string) => {
      context.helpers.push(s);
    }
  };

  return context;
}
