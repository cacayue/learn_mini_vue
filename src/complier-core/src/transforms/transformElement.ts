import { CREATE_ELEMENT_BLOCK, helperNameMap } from '../complierHelper';
import { NodeType } from '../node';

export function transformElement(node: any, context: any) {
  // body
  if (node.type === NodeType.ELEMENT) {
    context.helper(helperNameMap.get(CREATE_ELEMENT_BLOCK));
  }
}
