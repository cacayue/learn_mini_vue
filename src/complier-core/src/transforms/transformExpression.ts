import { NodeType } from '../node';

export function transformExpression(node: any) {
  if (node.type === NodeType.INTERPOLATION) {
    let processNode = processExpression(node.content);
    node.content = processNode;
  }
}
function processExpression(node: any) {
  let rawContent = node.content;
  node.content = `_ctx.${rawContent}`;
  return node;
}
