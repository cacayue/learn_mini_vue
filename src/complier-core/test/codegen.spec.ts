import { it, expect, describe } from 'vitest';
import { baseParse } from '../src/parse';
import { generateCode } from '../src/codegen';
import { transform } from '../src/transform';
import { transformExpression } from '../src/transforms/transformExpression';
import { transformElement } from '../src/transforms/transformElement';

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi');
    transform(ast);
    const { code } = generateCode(ast);

    expect(code).toMatchSnapshot();
  });

  it('interpolation', () => {
    const ast = baseParse('{{message}}');
    transform(ast, {
      nodeTransforms: [transformExpression]
    });
    const { code } = generateCode(ast);

    expect(code).toMatchSnapshot();
  });

  it('element', () => {
    const ast = baseParse('<div></div>');
    transform(ast, {
      nodeTransforms: [transformElement]
    });
    const { code } = generateCode(ast);

    expect(code).toMatchSnapshot();
  });
});
