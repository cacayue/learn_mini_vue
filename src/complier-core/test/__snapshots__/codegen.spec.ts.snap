// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`codegen > element 1`] = `
"import { createElementBlock as _createElementBlock } from 'vue'
return function render(_ctx, _cache) {return _createElementBlock(div)}"
`;

exports[`codegen > interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from 'vue'
return function render(_ctx, _cache) {return _toDisplayString(return _ctx.message)}"
`;

exports[`codegen > string 1`] = `"return function render(_ctx, _cache) {return hi}"`;
